Support ignoring failing dependencies (or dependencies altogether)

stdweird opened the pull request:
Give the admin advanced control over the dependencies, in particular to skip them altogether via `--nodeps --no-autodeps` or run the postdependencies best-effort via `--nodepsnoerrors`.

Another use case is in AII where  the initial `ncm-ncd --co spma` during `postinstall` can fail due to postdependencies of `spma` (e.g. `chkconfig`); which is OK in this specific case since no other components ran.

stdweird commented:
No ready to be merged, missing feedback and unittests

a) adds support for `--autodeps=0`; but `--autodeps` is now not supported anymore (argument is mandatory now). i guess this is not acceptable, so i can add a `--noautodeps` (however `--autodeps` itself is a useless option when the default is 1); and i would like to use `--nodepsnoautodeps` (i.e. `--nodeps --autodeps=0`), because that's probably what people expect from `--nodeps`.
b) `--nodepsnoerrors` is likely not the best option name ever. suggestions welcome
c) `--nodepsnoerrors` uses a slightly different interpretation of `dependency` 
d) related to c, `--nodeps` has an interesting issue/bug: if you have comp `X` with post dep `Y` and comp `Y` has predep `X`, then `ncm-ncd --co X --nodeps` will treat failure of `X` as ok, because `X` is a (pre)dependency of `Y`. (it's one of the TODO in the PR)

@ned21 @jouvin @jrha toughts?

jouvin commented:
My first remark(!!!) would be to ask the motivation/goal of this modification: could you put a few word in the description. As you said yourself, there are a lot of corner cases and evaluating if they are acceptable or not partly depend on the use case...

stdweird commented:
ah, yes.

initial / urgent use case: we have `chkconfig` as post dep of `spma`, and in AII there is a line
```bash
ncm-ncd --co spma || fail whatever
```
that does the initial `spma` run (followed by a  `ncm-ncd --co -all` on sucess)
with `ncm-systemd`, we have some services that fail initially (since they are not configured, no surprises), but this now makes the `ncm-ncd --co spma` fail, and thus the whole kickstart
(result: basically, we can't install EL7 nodes anymore with aii)

so my initial toughts were to support a `nodepsnoerrors` scenario, were the dependencies do run, but cause no major failure if they fail, since the main requested component `spma` does not fail (and in this specific case, it's ok that the deps are tried in a best effort way)

but since i cleaned up the code, i also fixed `--nodeps --no-autodeps`, so you can run components without any dependencies (hooray for faster debugging, as we have spma as predependency for everything, and spma can be slowish, even if it has nothing to do). that would also be a solution for our problem, but i like the concept of `nodepsnoerrors`.

stdweird commented:
All unittests added

jouvin commented on src/main/perl/ComponentProxy.pm:
```diff
@@ -336,7 +337,7 @@ sub _setDependencies {
         }
         $self->debug(2, "pre dependencies for component $self->{'NAME'}: ",
-                     ' '.join(',',@{$self->{PRE_DEPS}}));
+                     join(',',@{$self->{PRE_DEPS}}));
```
Is this formatting change really needed. It may breaks some scripts parsing this line. Looks gratuitous for me, doesn't really bring any value as a component name cannot contain a space.

stdweird replied:
not really needed no, but it's annoying and wrong (and this is debug level 2, not really error messages)

jouvin commented on src/main/perl/ComponentProxy.pm:
```diff
@@ -349,7 +350,7 @@ sub _setDependencies {
         }
         $self->debug(2, "post dependencies for component $self->{NAME}: ",
-                     ' '.join(',',@{$self->{POST_DEPS}}));
+                     join(',',@{$self->{POST_DEPS}}));
```
Idem...

jouvin commented on src/main/perl/ComponentProxyList.pm:
```diff
@@ -146,31 +172,42 @@ sub run_all_components
+                         # Convert errors in warnings
+                         # TODO do we set the state to failed?
+                         $self->warn("Errors from $name are downgraded to warnings (nodepsnoerrors is set). ",
+                                     "State is not cleared (as something went wrong), but also not set.");
```
Not sure what this line means... And the previous line looks more as a debug message than a warning message. What about a warning text like "$name execution failure ignored (nodepsnoerror is set)" and possibly a debug message with more information if neeeded.

stdweird replied:
i'll reformutlate the emssages, but do we
a. set the state to fail (as something wnet wrong)
b. clear the state (as we ignore the error)
c. nothing (and in that case i propose we log the fact that we doon't change the state)
?

jouvin commented on src/main/scripts/ncm-ncd:
```diff
@@ -119,13 +119,19 @@ sub app_options()
 
         {
-            NAME    => 'autodeps',
+            NAME    => 'autodeps!',
```
What is the exclamation point about?

stdweird replied:
in pure `Getopt::Long`, it means the a negation of form `--no-<optioname>` or `--no<optionname>` can disable it (but it seems to already work somehow). i wasn't aware of it and i want to make it explicit

jouvin commented on src/main/scripts/ncm-ncd.pod:
```diff
@@ -107,11 +107,30 @@ use separate (per component) log files in log directory
+'--autodeps', if any).
+
+(If you do not care about dependencies and just want to avoid errors,
+you can also try to use '--nodeps --no-autodeps').
```
Should make clear that this must not be used in normal circumstances and that this is mainly for testing.

stdweird replied:
shall i also add it to `--nodeps`, `--no-autodeps` and `--allowbrokencomps`?

stdweird replied:
i reordered the pod and moved anything potential dangerous to the "Advanced Options" and added a `Use with care` message

jouvin commented:
I understand the use case and I think that I agree that your proposed solution is good. I still have 2 remarks:

- I find `nodepsnoerror` a very cryptic option name... Why not `ignore-dep-failures` or something like this.
- I have mixed feeling with this option applying both to pre and post dependencies. In the use case you mention, this is clearly post dependency failures that you want to ignore. If a pre dependency was failing, it would be discussable if you could really ignore it (except for testing purpose covered by `--nodeps --noautodeps`...

stdweird commented:
@jouvin wrt the optionname, i don't mind changing it, just simple search and replace
wrt only-pre dependecies, yes, using `--nodeps --noautodeps` is now a possibility, but i don't see why the difference matters. i consider postdependencies as important as predependencies (and if i decide to not care as much, i also don't care equally :smile: ); but i sortof see your point. at least, if only post dependecies are affected, the optionname has to make this clear.

jouvin commented:
I also consider predependencies and postdependencies equally important. But their impact on the run of a configuration module is different: if a predependency fails, there is a non-zero risk that the configuration module will not be able to run successfully and it doesn't really make sense running the configuration module (except if you know what you are doing and in this case `--nodeps --noauthdeps` will do the job). Where a failed postdependency doesn't affect the ability to run a configuration module, only it will probably make the configuration incomplete/inconsistent. This is something that makes sense at installation time and having an option that allows only this makes it really different from `--nodeps --noautodeps` in my opinion.

stdweird commented:
@jouvin wrt to post-dependencies only, this is not that easy.
see https://github.com/quattor/ncm-ncd/pull/44/files#diff-daa11034fa1ae434ee5db3f1c269ee57R145
at that moment, it is hard to figure out if a that component is not a predependency of any of the requested components; i would need to create the a list of reverse topological sorted components first.

ned21 commented on src/main/scripts/ncm-ncd:
```diff
@@ -183,6 +155,41 @@ sub app_options()
+        },
+
+        {
+            NAME    => 'nodepsnoerrors',
```
How about --ignore-errors-from-dependencies as a name for this option?

ned21 replied:
An explanation of this option also needs to be added to the ncm-ncd.pod file.

stdweird replied:
already in the pod

ned21 commented on src/main/scripts/ncm-ncd:
```diff
@@ -183,6 +155,41 @@ sub app_options()
+
+        {
+            NAME    => 'nodepsnoerrors',
+            HELP    => 'errors from broken (pre/post) dependencies in configure are downgraded to warnings (implies --nodeps --autodeps). Use with care.',
```
What is the meaning of "broken" in this context? Running to completion but with ERRORs or some sort of code exception? (i.e. code is buggy)

stdweird replied:
i replaced `broken` with `failing` to inidicate that they run to completion, but with errors

ned21 commented:
How about an option to ignore components by name? e.g. --ignore-errors-from $name. That allows it to ignored whether it appears in either pre or post, and restricts the admin to ignoring only those components known to fail and if others unexpectedly fail then ncm-ncd behaves as today?

stdweird commented:
@ned21 wrt `--ignore-errors-from $name`, why not also/instead support `--skip $name` for non-all selection of components. (but can you open a new issue for that). looks like we need some matrix of case we want to handle, and see what options are rqeuired and if new ones need to be added/extended.

@jouvin, @jrha i'll go with `ignore-errors-from-dependencies`

jrha commented:
I like `ignore-errors-from-dependencies`, it's nice and clear :grinning:

stdweird commented:
Was there anything left to address here?

ned21 commented:
LGTM
